File: src/de/umr/fixcon/graphFunctions/standardFunctions/IsTreeFunction.java
package de.umr.fixcon.graphFunctions.standardFunctions;

import com.google.common.graph.Graph;
import com.google.common.graph.Graphs;
import de.umr.fixcon.graphFunctions.DecisionProblem;
import de.umr.fixcon.graphFunctions.GraphFunction;

import static com.google.common.base.Preconditions.checkArgument;

public class IsTreeFunction extends DecisionProblem implements GraphFunction {

    @Override
    public boolean isEdgeMonotone() {
        return false;
    }

    @Override
    public double apply(final Graph<Integer> g, final int... parameters) {
        checkArgument(!g.nodes().isEmpty());
        boolean isConnected = Graphs.reachableNodes(g, g.nodes().iterator().next()).equals(g.nodes());
        boolean isAcyclic = !Graphs.hasCycle(g);
        return (isAcyclic && isConnected) ? 1.0 : 0.0;
    }
}

File: src/de/umr/fixcon/graphFunctions/standardFunctions/IsTreeFunction.kt
package de.umr.fixcon.graphFunctions.standardFunctions

import de.umr.fixcon.graphFunctions.GraphFunction
import org.jgrapht.Graph
import org.jgrapht.GraphTests
import org.jgrapht.GraphTests.isTree
import org.jgrapht.graph.DefaultEdge

/**
 * This function returns 1 (indicator for **True**) iff the graph is a tree (connected and acyclic).
 */
object IsTreeFunction : GraphFunction {
    override val isEdgeMonotone: Boolean = false

    override fun apply(g: Graph<Int, DefaultEdge>, args: List<Int>) = if (isTree(g)) 1.0 else 0.0
}